include/spline/de_casteljau_subdivide.hpp

```diff
@@ -3,116 +3,129 @@
 /// @brief The Spline library
-namespace spline
-{
-/// @brief De Casteljau's algorithm with subdivision
-/// @tparam InputIt Input iterator.
-/// @tparam OutputIt Output iterator.
-/// @tparam Real Real number type (typically float/double/long double).
-/// @tparam MultiplicationOp binary operator accepting
-///         InputIt::value_type and Real, returning an InputIt::value_type.
-/// @tparam AdditionOp binary operator accepting two InputIt::value_type
-///         and returning a InputIt::value_type.
-/// @param[in] first Iterator to first element of input range.
-/// @param[in] last Iterator to one past the last element of the range.
-/// @param[in] t Point at which to compute the polynomial value.
-/// @param[in] mul Vector-Scalar multiplication function object.
-/// @param[in] add Vector-Vector addition function object.
-/// @param[out] d_first Beginning of destination range.
-/// @return Output iterator to the element past the last element written.
-/// @pre @p d_first must point to a destination range where it is possible to
-///      write  2 * std::distance(first, last) - 1 elements.
-template <class InputIt, class OutputIt, class Real, class MultiplicationOp,
-          class AdditionOp>
-constexpr auto de_casteljau_subdivide(InputIt first, InputIt last,
-                                      OutputIt d_first, Real t,
-                                      MultiplicationOp mul, AdditionOp add)
-    -> OutputIt
-{
-    auto const num_coefficients = last - first;
-    if (num_coefficients <= 0)  // Empty or non-sensical input range.
+namespace spline {
+
+namespace detail {
+struct de_casteljau_subdivide_fn final {
+    /// @brief De Casteljau's algorithm with subdivision
+    /// @tparam InputIt Input iterator.
+    /// @tparam OutputIt Output iterator.
+    /// @tparam Scalar Scalar number type (typically float/double/long double).
+    /// @tparam MultiplicationOp binary operator accepting
+    ///         InputIt::value_type and Scalar, returning an
+    ///         InputIt::value_type.
+    /// @tparam AdditionOp binary operator accepting two InputIt::value_type
+    ///         and returning a InputIt::value_type.
+    /// @param[in] first Iterator to first element of input range.
+    /// @param[in] last Iterator to one past the last element of the range.
+    /// @param[in] t Point at which to compute the polynomial value.
+    /// @param[in] mul Vector-Scalar multiplication function object.
+    /// @param[in] add Vector-Vector addition function object.
+    /// @param[out] d_first Beginning of destination range.
+    /// @return Output iterator to the element past the last element written.
+    /// @pre @p d_first must point to a destination range where it is possible
+    /// to write  2 * std::distance(first, last) - 1 elements.
+    template <class InputIt, class OutputIt, class Scalar,
+              class MultiplicationOp, class AdditionOp>
+    constexpr auto operator()(InputIt first, InputIt last, OutputIt d_first,
```

Comment:
instead of moving the implementation into a struct, what do you think of wrapping the free function so that clients which prefer not to (or can't) use some of the STL headers don't have to?

Reply:
As an example, here is what it means to pull in <functional>:
https://github.com/llvm/llvm-project/blob/main/libcxx/include/functional#L509

Compile time aside, it would be nice if it were still possible to use de casteljau without needing to pull in all these headers.

Reply:
Do you mean avoiding <type_traits>, <functional>, <iterator>?

These are used to define the constraints, so we could probably do the following:
Split into `impl/de_casteljau_subdivide.hpp` and `de_casteljau_subdivide.hpp` where

`impl/de_casteljau_subdivide.hpp`: free function, no constraints
`de_casteljau_subdivide.hpp`: function object forwarding to the free function, with constraints

Reply:
FYI, I use `<numeric>` in a follow up to simply (imo) the algorithm implementation.

Reply:
I moved this into an `impl` namespace, but I imagine there to be two top level include files:
- `spline.hpp`: with constraints and nice default function objects
- `spline_core.hpp`: just the algorithms

Reply:
Any reason why Mul comes before Add? Seems like the other way is more common in literature.